core/src/apps/zcash/sig_hasher.py

```diff
@@ -0,0 +1,187 @@
+        write_bytes_prefixed(self.scriptpubkeys, script_pubkey)  # (see S.2d scriptpubkeys_sig_digest)
+        write_uint32(self.sequence, txi.sequence)                # (see S.2e sequence_sig_digest)
+
+    def add_output(self, txo: TxOutput, script_pubkey: bytes) -> None:
+        self.empty = False
+        write_tx_output(self.outputs, txo, script_pubkey)        # (see S.2f outputs_sig_digest)
+
+    def digest(self):
+        """Returns `T.2: transparent_digest` field for txid computation."""
+        h = HashWriter(blake2b(outlen=32, personal=b'ZTxIdTranspaHash'))
+
+        if not self.empty:
+            write_hash(h, self.prevouts.get_digest())  # T.2a: prevouts_digest (32-byte hash)
+            write_hash(h, self.sequence.get_digest())  # T.2b: sequence_digest (32-byte hash)
+            write_hash(h, self.outputs.get_digest())   # T.2c: outputs_digest  (32-byte hash)
+
+        return h.get_digest()
+
+    def sig_digest(self, txin_sig_digest: bytes | None):
+        """Returns `S.2: transparent_sig_digest` field for signature digest computation."""
+        if self.empty:
+            assert txin_sig_digest is None
+            return self.digest()
+
+        if txin_sig_digest is None:
+            txin_sig_digest = blake2b(outlen=32, personal=b'Zcash___TxInHash').digest()
+
+        h = HashWriter(blake2b(outlen=32, personal=b'ZTxIdTranspaHash'))
+
+        # only SIGHASH_ALL is supported in Trezor
+        write_uint8(h, SigHashType.SIGHASH_ALL)         # S.2a: hash_type                (1 byte)
+        write_hash(h, self.prevouts.get_digest())       # S.2b: prevouts_sig_digest      (32-byte hash)
+        write_hash(h, self.amounts.get_digest())        # S.2c: amounts_sig_digest       (32-byte hash)
+        write_hash(h, self.scriptpubkeys.get_digest())  # S.2d: scriptpubkeys_sig_digest (32-byte hash)
+        write_hash(h, self.sequence.get_digest())       # S.2e: sequence_sig_digest      (32-byte hash)
+        write_hash(h, self.outputs.get_digest())        # S.2f: outputs_sig_digest       (32-byte hash)
+        write_hash(h, txin_sig_digest)                  # S.2g: txin_sig_digest          (32-byte hash)
+
+        return h.get_digest()
+
+
+def get_txin_sig_digest(
+    txi: TxInput,
+    public_keys: Sequence[bytes | memoryview],
+    threshold: int,
+    coin: CoinInfo,
+    sighash_type: int,
+):
+    """Returns `S.2g: txin_sig_digest` field for signature digest computation."""
+    assert sighash_type == SigHashType.SIGHASH_ALL
+    h = HashWriter(blake2b(outlen=32, personal=b'Zcash___TxInHash'))
+
+    write_prevout(h, txi)                       # S.2g.i:   prevout      (field encoding)
+    write_uint64(h, txi.amount)                 # S.2g.ii:  value        (8-byte signed little-endian)
+    scripts.write_bip143_script_code_prefixed(  # S.2g.iii: scriptPubKey (field encoding)
+        h, txi, public_keys, threshold, coin
+    )
```

Comment:
I think in the end the Zcash people decided on really using the scriptPubKey, not the scriptCode, didn't they?

Reply:
You are right. They decided to use *scriptPubKey* according to [this comment](https://github.com/zcash/zips/pull/577#discussion_r781645348). 

However notice that calling `write_bip143_script_code_prefixed` has currently only two possible results:
- P2PKH scriptPubKey for multiple possible script types
- multisig scriptPubKey for multisig

The function always return a scriptPubKey so it could be renamed to `write_bip143_script_pubkey_prefixed`. Or even to `write_script_pubkey_prefixed` since it is currently used only in Zcash and it is not further related to the BIP-143.

Or preferably just to `write_script_pubkey` with `prefixed: bool = true` argument for sake of universality.

Reply:
https://github.com/trezor/trezor-firmware/pull/2166/commits/060aa20c4d7490f209b520230ea2122f93dd4dcb